docs/module-writing-basics.md

```diff
@@ -0,0 +1,195 @@
+
+### KeepThreadedContainersInState()
+
+Specific to `ThreadAwareModule` this method returns True in the base class. If set to false by the child class, containers used for parallel processing will be popped from the state.
+
+### SetUp()
+
+Called by the orchestration only once, with parameters as defined by the recipe file. 
+
+### PreProcess()
+
+Specific to `ThreadAwareModule`, this method is called exactly once by the orchestration between `SetUp()` and `Process()`.
+
+### Process()
+
+This method performs the bulk of the processing action. This method is called for each module in parallel, based on dependencies as outlines by the recipe file. 
+
+For a `BaseModule` no parameters will be passed in, but for a `ThreadAwareModule` the method will be called with a single container of the type specified in `GetThreadOnContainerType()`
+
+### PostProcess()
+
+Specific to `ThreadAwareModule`, this method is called exactly once by the orchestration after `Process()`.
+
+## 1 - Parallel Processing Helpers
+
+Three of the methods outlined above are used to tell the orchestration how to handle parallel processing.
+
+The `GetThreadOnContainerType()` method is used to identify which container is used for threading. That is, this is the container type that will get passed to `Process(container)` one at a time. Since we're operating on local filesystem files, we can use the existing container, `containers.File`.
+
+```python
+	@staticmethod
+	def GetThreadOnContainerType():
+		return containers.File
+```
+
+Second of these methods is `GetThreadPoolSize()` which tells the orchestration how many threads this module supports. Since we are doing local processing, we will set the number of threads to be based on the number of CPUs in the machine doing the work - halved to ensure we're not being greedy on local resources.
+
+```python
+	def GetThreadPoolSize(self):
+		count = math.floor(os.cpu_count() / 2)
+		return 1 if count < 1 else count
+```
+
+Finally, we want to operate on file containers, and only keep those file containers in the state if the hash is matched. We will do that by telling the orchestration to pop the containers when passed to `Process()` and add the matches back.
+
+```python
+	@staticmethod
+	def KeepThreadedContainersInState():
+		return False
+```
+
+## 2 - Initialisation
+
+Two methods are used as part of module initialisation: The python class `__init__()` and DFTimewolf's `SetUp()` method.
+
+\_\_init__ is simple enough, we only need to declare the class members:
+
+```python
+	def __init__(self, state, name, critical):
+		super(FileRegexModule, self).__init__(state, name=name, critical=critical)
+		self.hashes = []
+		self.destination_dir = ''
+```
+
+`SetUp()` is called by the orchestration with parameters as defined by the recipe file. Our module is going to receive two parameters: a comma separated list of files and a comma separated list of hashes. 
```

Comment:
Might be worth specifying that they're getting comma-separated strings instead of lists because they're supposed to be passed in by humans through the CLI (and not directly from a recipe)

Reply:
Done.